yt_dlp/extractor/nebula.py

```diff
@@ -1,281 +1,385 @@
+        try:
+            return self._download_json(*args, **kwargs)
+        except ExtractorError as e:
+            if not isinstance(e.cause, urllib.error.HTTPError) or e.cause.status not in (401, 403):
+                raise
+            self.to_screen(f'Reautherizing with Nebula and retrying, '
+                           f'because last API call resulted in error {e.cause.status}')
+            self._real_initialize()
+            if self._token:
+                kwargs.setdefault('headers', {})['Authorization'] = f'Bearer {self._token}'
+            return self._download_json(*args, **kwargs)
 
-        def inner_call():
-            authorization = f'Token {self._nebula_api_token}' if auth_type == 'api' else f'Bearer {self._nebula_bearer_token}'
-            return self._download_json(
-                url, video_id, note=note, headers={'Authorization': authorization},
-                data=b'' if method == 'POST' else None)
+    def _real_initialize(self):
+        if not self._api_token:
+            self._api_token = try_call(
+                lambda: self._get_cookies('https://nebula.tv')['nebula_auth.apiToken'].value)
+        self._token = self._download_json(
+            'https://users.api.nebula.app/api/v1/authorization/', None,
+            headers={'Authorization': f'Token {self._api_token}'} if self._api_token else None,
+            note='Authorizing to Nebula', data=b'')['token']
 
+    def _extract_formats(self, content_id, slug, fatal=False):
+        api_path = 'lessons' if content_id.startswith('lesson:') else 'video_episodes'
         try:
-            return inner_call()
-        except ExtractorError as exc:
-            # if 401 or 403, attempt credential re-auth and retry
-            if exc.cause and isinstance(exc.cause, HTTPError) and exc.cause.status in (401, 403):
-                self.to_screen(f'Reauthenticating to Nebula and retrying, because last {auth_type} call resulted in error {exc.cause.code}')
-                self._perform_login()
-                return inner_call()
-            else:
+            fmts, subs = self._extract_m3u8_formats_and_subtitles(
+                f'https://content.api.nebula.app/{api_path}/{content_id}/manifest.m3u8',
+                slug, 'mp4', query={
+                    'token': self._token,
+                    'app_version': '23.10.0',
+                    'platform': 'ios',
+                })
+        except ExtractorError as e:
+            if isinstance(e.cause, urllib.error.HTTPError) and e.cause.status == 401:
+                self.raise_login_required('This video is only available for users with an active subscription')
+            if fatal or not isinstance(e.cause, urllib.error.HTTPError) or e.cause.status != 403:
                 raise
+            self.to_screen('Reautherizing with Nebula and retrying, because fetching video resulted in error')
+            self._real_initialize()
+            return self._extract_formats(content_id, slug, fatal=True)
```

Comment:
instead of adding the `fatal` param and recursing, we could use a retry loop, e.g.
```py
    def _extract_formats(self, content_id, slug):
        api_path = 'lessons' if content_id.startswith('lesson:') else 'video_episodes'
        for retry in (False, True):
            try:
                fmts, subs = self._extract_m3u8_formats_and_subtitles(
                    f'https://content.api.nebula.app/{api_path}/{content_id}/manifest.m3u8',
                    slug, 'mp4', query={
                        'token': self._token,
                        'app_version': '23.10.0',
                        'platform': 'ios',
                    })
                return {'formats': fmts, 'subtitles': subs}
            except ExtractorError as e:
                if isinstance(e.cause, HTTPError) and e.cause.status == 401:
                    self.raise_login_required()
                if not retry and isinstance(e.cause, HTTPError) and e.cause.status == 403:
                    self.to_screen('Reauthorizing with Nebula and retrying because format extraction failed')
                    self._real_initialize()
                    continue
                raise
```